### ISSUE TITLE
Occasional stale sessions that can’t be refreshed

### ISSUE THREAD

**Issue by ryandawsonuk** [arrikto/oidc-authservice#2]
We’ve been using this in kubeflow and initially everything works fine. But after you’ve been running roughly more than a week then your session can seemingly be expired and you get 403s. This might kinda make sense but there’s then no way in kubeflow to start a new session. You can’t even logout as every page gets blocked. 

When this happens the authservice pod doesn’t log anything for each 403. But it is the authservice rejecting requests as we can fix it by killing that Pod and letting it come back up again. When it comes back up we can login as normal. 

Naturally this is quite hard to replicate as you have to run for a long time before you hit it. We didn’t encounter this previously with the old ambassador-oidc version.

**Comment by yanniszark** [comment]
@ryandawsonuk thanks for reporting!
This is a peculiar scenario.

You mentioned that you suspect the oidc-authservice replies with 403 because the session is expired.
However, restarting the oidc-authservice Pod should have no effect on the expiry of the token.
As such, I suspect it is a different issue.

A thing I should mention is that getting 403 can also result from the Gateway not being able to contact the AuthService.
Maybe it gets stuck?

I will make sure to add more logging that can aid in debugging.
In the meantime, if you see this again, can you check if the network request is arriving at the gateway? ( with tcpdump, strace in the Pod)

**Comment by SachinVarghese** [comment]
@yanniszark You are right. The requests are not reaching the auth service! I do not have any logs at the moment, will attach gateway trace if this happens again.

**Comment by ryandawsonuk** [comment]
Any ideas on what more we could do to investigate? It wasn't happening prior to 0.7.0. Maybe there were also istio config version or changes around that time? It's currently looking like a big issue for us.

**Comment by ryandawsonuk** [comment]
I guess you mean to do tcpdump or strace on the istio gateway pods? I can't [exec into the authservice container ](https://nmrony.info/ssh-into-kubernetes/) and neither strace nor tcpdump are on the path. I get `command terminated with exit code 126` for each so all we can see are the logs there.

**Comment by webD97** [comment]
I can confirm this behavior but unfortunately I have no additional information for you.

**Comment by me** [comment]
I'm also seeing the same behaviour. Exec'ing into the authservice, I can see that the service is still somewhat responsive to a bad request:
```
~ $ nc localhost 8080
foo
HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Connection: close

400 Bad Request
```

but a proper request just hangs indefinitely

```
~ $ nc localhost 8080
GET / HTTP/1.1
host: localhost:8080

^Cpunt!
```

On the ingressgateway side, tcpdump shows the request being made and no response.

**Comment by yanniszark** [comment]
Thanks for reporting this.
It seems like a lot of people are encountering this behavior.
I suspect the culprit is the BoltDB library we use for the session storage (yosssi/boltstore).
I have updated it to the latest version, so hopefully that will fix it.

Otherwise, I will try to reproduce and attach a delve debugger to the process.
I have also prepared a new image from master: `gcr.io/arrikto/kubeflow/oidc-authservice:eda1ee8`
Feel free to try it out and report back.

Finally, I opened a PR in kubeflow/manifests to get this fix in: https://github.com/kubeflow/manifests/pull/633

**Comment by ryandawsonuk** [comment]
Closing this as the bolt version change seems to fix it. Not been able to reproduce on that version. Can reopen if we hit it again.

**Comment by ittus** [comment]
I encountered the same issue with version https://github.com/arrikto/oidc-authservice/commit/e2364397aaf1a8119aa649989f0de87276f58cbc

**Comment by athamark** [comment]
Hello @ittus, do you encounter this problem for the latest available AuthService image [`ae92e86`](https://console.cloud.google.com/gcr/images/arrikto/global/oidc-authservice@sha256:022f60055c4d120675d5566963cb550c0980926c338db9eeac2cb76c23d1c781/details)? 
Note that we have started pushing our AuthService images to the `gcr.io/arrikto/oidc-authservice` registry. We have standardize on this decision and therefore all the upcoming official AuthService images will be served there. Here is a link to the new registry: https://console.cloud.google.com/gcr/images/arrikto/GLOBAL/oidc-authservice

> If you wish to use any older AuthService images you can checkout the `gcr.io/arrikto/kubeflow/oidc-authservice` registry. In there, you will find images that date back to June of 2019. Here is a link to the old registry: https://console.cloud.google.com/gcr/images/arrikto/GLOBAL/kubeflow/oidc-authservice

**Comment by ittus** [comment]
Hi @athamark ,
We had the issue when using the image in https://console.cloud.google.com/gcr/images/arrikto/GLOBAL/kubeflow/oidc-authservice with https://github.com/arrikto/oidc-authservice/commit/e2364397aaf1a8119aa649989f0de87276f58cbc tag.

We did some metrics logging, and figured out that the boltdb size increase a lot. After a while, it makes the pod hang.
Our current solution is replacing boltdb with redis, which has expire feature work pretty nice.
